pymarsys/connections.py

```diff
@@ -0,0 +1,143 @@
+from abc import ABC, abstractmethod
+import base64
+import datetime
+import hashlib
+import json
+import os
+import uuid
+
+import requests
+from aiohttp import ClientSession
+
+EMARSYS_URI = 'https://api.emarsys.net/'
+
+
+class ApiCallError(Exception):
+    pass
+
+
+class BaseConnection(ABC):
+    """
+    Any connection used to instantiate an Emarsys object or an object inherited
+     from BaseEndpoint should inherit from this class.
+    this class.
+    """
+    @abstractmethod
+    def __init__(self, username, secret, uri):
+        self.username = username
+        self.secret = secret
+        self.uri = uri
+
+    def build_authentication_variables(self):
+        """
+        Build the authentication variables Emarsys' authentication system
+        asks for.
+        :return: nonce, created, password_digest.
+        """
+        nonce = uuid.uuid4().hex
+        created = datetime.datetime.utcnow().strftime(
+            '%Y-%m-%dT%H:%M:%S+00:00'
+        )
+        sha1 = hashlib.sha1(
+            str.encode(nonce + created + self.secret)
+        ).hexdigest()
+        password_digest = bytes.decode(base64.b64encode(str.encode(sha1)))
+        return nonce, created, password_digest
+
+    def build_headers(self, other_http_headers=None):
```

Comment:
[personal taste]: i would rather define this function as a staticmethod because it is not especially dependent of many class attributes, and using a staticmethod forces oneself to define function parameters